src/Common/src/Common.Expression/Spring/Ast/SpelNode.cs

```diff
@@ -140,95 +139,191 @@ protected internal virtual IValueRef GetValueRef(ExpressionState state)
+            {
+                // This would be unnecessary in the case of subtyping (e.g. method takes Number but Integer passed in)
+                CodeFlow.InsertCastClass(gen, paramDesc);
+            }
+
+            cf.ExitCompilationScope();
+        }
+
+        protected static OpCode GetLdElemInsn(Type arrElemType)
+        {
+            if (arrElemType == typeof(sbyte))
+            {
+                return OpCodes.Ldelem_I1;
+            }
+
+            if (arrElemType == typeof(byte))
+            {
+                return OpCodes.Ldelem_U1;
+            }
+
+            if (arrElemType == typeof(short) || arrElemType == typeof(char))
+            {
+                return OpCodes.Ldelem_I2;
+            }
+
+            if (arrElemType == typeof(ushort))
+            {
+                return OpCodes.Ldelem_U2;
+            }
+
+            if (arrElemType == typeof(int))
+            {
+                return OpCodes.Ldelem_I4;
+            }
+
+            if (arrElemType == typeof(uint))
+            {
+                return OpCodes.Ldelem_U4;
+            }
+
+            if (arrElemType == typeof(long))
+            {
+                return OpCodes.Ldelem_I8;
+            }
+
+            if (arrElemType == typeof(ulong))
+            {
+                return OpCodes.Ldelem_I8;
+            }
+
+            if (arrElemType == typeof(float))
+            {
+                return OpCodes.Ldelem_R4;
+            }
+
+            if (arrElemType == typeof(double))
+            {
+                return OpCodes.Ldelem_R8;
+            }
+
+            if (arrElemType == typeof(IntPtr) || arrElemType == typeof(UIntPtr))
+            {
+                return OpCodes.Ldelem_I;
+            }
+
+            return OpCodes.Ldelem_Ref;
+        }
+
+        protected static OpCode GetStelemInsn(Type arrElemType)
+        {
+            if (arrElemType == typeof(sbyte) || arrElemType == typeof(byte) || arrElemType == typeof(bool))
+            {
+                return OpCodes.Stelem_I1;
+            }
+
+            if (arrElemType == typeof(short) || arrElemType == typeof(ushort) || arrElemType == typeof(char))
+            {
+                return OpCodes.Stelem_I2;
+            }
+
+            if (arrElemType == typeof(int) || arrElemType == typeof(uint))
+            {
+                return OpCodes.Stelem_I4;
+            }
+
+            if (arrElemType == typeof(long) || arrElemType == typeof(ulong))
+            {
+                return OpCodes.Stelem_I8;
+            }
+
+            if (arrElemType == typeof(float))
+            {
+                return OpCodes.Stelem_R4;
+            }
+
+            if (arrElemType == typeof(double))
+            {
+                return OpCodes.Stelem_R8;
+            }
+
+            if (arrElemType == typeof(IntPtr) || arrElemType == typeof(UIntPtr))
+            {
+                return OpCodes.Stelem_I;
+            }
+
+            return OpCodes.Stelem_Ref;
```

Comment:
```suggestion
        protected static OpCode GetStelemInsn(object arrElemType)
        {
            return arrElemType switch
            {
                sbyte or byte or bool => OpCodes.Stelem_I1,
                short or ushort or char => OpCodes.Stelem_I2,
                int or uint => OpCodes.Stelem_I4,
                long or ulong=> OpCodes.Stelem_I8,
                float => OpCodes.Stelem_R4,
                double => OpCodes.Stelem_R8,
                IntPtr or UIntPtr => OpCodes.Stelem_I,
                _ => OpCodes.Stelem_Ref
            };
```
This change also creates test failures, likely for the same reason as the above change... If we can make one work we can make them both work and I think it reads a lot easier as a result, but probably isn't worth spending a whole lot of time on it if it can't be fixed quickly

Reply:
Pushed changes